Groups

Bilb opened the pull request:
Adds a metagroup wrapper to manage the state of groups with libsession.
This metagroupwrapper has a ref to the corresponding info, members and keys relevant to that group.

Bilb committed:
feat: add working GroupInfo constructor

Bilb committed:
feat: add createGroup getGroup and destroyGroup

Bilb committed:
feat: wip of info and members group

Bilb committed:
Merge remote-tracking branch 'origin/main' into groups

Bilb committed:
feat: wip of groupkeys wrapper

Bilb committed:
feat: add metagroup wrapper compiling with some methods only

Bilb committed:
feat: add start of metagroup wrapper+js bindings

Bilb committed:
feat: add meta_group wrapper functions exported to JS

Bilb committed:
broken state after metagroup dump and load

Bilb committed:
chore: removed useless metabase

Bilb committed:
feat: first working version with metagroup wrapper

Bilb committed:
feat: add groups set, erase and getAll from usergroups

Bilb committed:
feat: add metaMerge for group wrapper

Bilb committed:
feat: add a type for 05 pubkeys and blinded ones

Bilb committed:
fix: do not confirmKeys when pushed. wait for poll event

Bilb committed:
feat: add encrypt/decrypt and expose make_dump

Bilb committed:
fix: compile issues and bump libsession-util

Bilb committed:
fix: linking error and remove groupKeys function for now

Bilb committed:
chore: pull protobuf branch

Bilb committed:
feat: add makeDump function to configBase calls too

Bilb committed:
feat: libsession merge returns vector of hashes merged now

Bilb commented on src/user_groups_config.cpp:
```diff
@@ -56,6 +57,22 @@ struct toJs_impl<legacy_group_info> {
+        obj["priority"] = toJs(env, info.priority);
+        obj["joinedAtSeconds"] = toJs(env, info.joined_at);
+        obj["name"] = toJs(env, info.name);
+        // TODO kicked and setKicked todo
```
I think this is part of another chunk

yougotwill replied:
Cool. feel free to resolve

Bilb committed:
fix: no point of having a shared_ptr for the meta_group

Bilb committed:
remove: removed unused configsig file

yougotwill commented on src/groups/meta_group_wrapper.hpp:
```diff
@@ -0,0 +1,317 @@
+#include "../profile_pic.hpp"
+#include "../utilities.hpp"
+#include "./meta_group.hpp"
+#include "oxenc/bt_producer.h"
```
Should this import not go under `# include <napi.h>` from what I saw in `base_config.hpp`? 

Is there a preferred import order for this project?

Bilb replied:
I just have auto format import on save enabled, and it keeps putting oxenc at the end, as `o > .` in ascii

yougotwill replied:
Ah okay

yougotwill commented on src/groups/meta_group_wrapper.hpp:
```diff
@@ -0,0 +1,317 @@
+
+    Napi::Value needsPush(const Napi::CallbackInfo& info);
+
+    Napi::Value push(const Napi::CallbackInfo& info) {
```
Is there a specific reason why the implementation is here and not in the cpp file?

yougotwill replied:
Same with the other functions below.

Bilb replied:
ok, I moved all of them, it's nice a pretty now. 
![image](https://github.com/oxen-io/libsession-util-nodejs/assets/1544279/1e21dd01-422d-476c-8595-3e61081d028c)

yougotwill commented on src/groups/meta_group_wrapper.hpp:
```diff
@@ -0,0 +1,317 @@
+
+            oxenc::bt_dict_producer combined;
+
+            // NB: the keys have to be in ascii-sorted order:
```
Can you please change NB to NOTE? 🙏

yougotwill replied:
This makes it similar to session-desktop

Bilb replied:
fixed

yougotwill commented on src/meta/meta_base_wrapper.hpp:
```diff
@@ -0,0 +1,109 @@
+    virtual ~MetaBaseWrapper() = default;
+
+    template <typename T, std::enable_if_t<is_derived_napi_wrapper<T>, int> = 0>
+    static void NoBaseClassInitHelper(
```
What exactly does this do?

Bilb replied:
explaining by chat

yougotwill commented on src/meta/meta_base_wrapper.hpp:
```diff
@@ -0,0 +1,109 @@
+            }
+
+            // Note, we keep shared_ptr for those as the Keys one need a reference to Members and
+            // Info on his own currently.
```
`on it's own`?

Bilb replied:
fixed with `on its own`

yougotwill commented on src/user_groups_config.cpp:
```diff
@@ -237,4 +262,74 @@ Napi::Value UserGroupsWrapper::eraseLegacyGroup(const Napi::CallbackInfo& info)
+        auto group_info = config.get_or_construct_group(groupPk);
+
+        if (auto priority =
+                    maybeNonemptyInt(obj.Get("priority"), "UserGroupsWrapper::setGroup priority")) {
```
maybeNonemptyInt -> maybeNonEmptyInt ?

Bilb replied:
There are a bunch of those, do you think I rename all of them ?
maybeNonemptyInt maybeNonemptyString  maybeNonemptyBuffer

Bilb replied:
we'll keep this one opened, I'd like to rename them once I rebased disappearing messages

yougotwill commented on src/user_groups_config.cpp:
```diff
@@ -237,4 +262,74 @@ Napi::Value UserGroupsWrapper::eraseLegacyGroup(const Napi::CallbackInfo& info)
+            group_info.priority = toPriority(obj.Get("priority"), group_info.priority);
+        }
+
+        if (auto joinedAtSeconds = maybeNonemptyInt(
```
This is a pretty clever convention (assigning the member in the if statement and then passing it to the object if true). Might try in session-desktop in future

Bilb replied:
yeah that is pretty neat. We actually don't have, but also don't need that kind of thing in typescript.
All of this does, is make sure that in the if statement, joinedAtSeconds is not null, so we can use it. i.e. it creates another varialbe to make sure that the code inside the if is aware that joinedAtSeconds is not null.

but in typescript, checkin the variable itself makes sure it is not null already in the if statement.
like this

```ts
const joinedAtSeconds: number | undefined;
if(joinedAtSeconds) {
   // here joinedAtSeconds is guaranteed to be non undefined/null (and also not 0)
}


```

yougotwill commented on src/utilities.hpp:
```diff
@@ -93,6 +97,13 @@ template <>
 struct toJs_impl<bool> {
     auto operator()(const Napi::Env& env, bool b) const { return Napi::Boolean::New(env, b); }
 };
+template <>
```
new line above please 🙏

Bilb replied:
fixed

yougotwill commented on src/utilities.hpp:
```diff
@@ -123,6 +134,17 @@ struct toJs_impl<std::vector<T>> {
+            arr[i] = toJs(env, as_array[i]);
+        return arr;
+    }
+};
```
new line

Bilb replied:
fixed

yougotwill commented on src/utilities.hpp:
```diff
@@ -123,6 +134,17 @@ struct toJs_impl<std::vector<T>> {
     }
 };
 template <typename T>
```
Add here

Bilb replied:
fixed for the whole file

yougotwill commented on types/groups/groupinfo.d.ts:
```diff
@@ -0,0 +1,15 @@
+declare module 'libsession_util_nodejs' {
+  /**
+   *
+   * Group Wrapper Logics
```
I guess this comment is redundant now? You don't make the same one in groupkeys.d.ts. But as you want 👍

yougotwill replied:
As I see you make it in later types so I guess groupkeys.d.ts is missing the comment 🤷

Bilb replied:
so I removed that specific comment. Each file has a // xxx related methods.
Added one for groupkeys yeah

yougotwill commented on types/groups/metagroup.d.ts:
```diff
@@ -0,0 +1,129 @@
+        groupKeys,
+        groupMember,
+      }: {
+        groupInfo: Array<MergeSingle> | null;
```
I wonder if you can group these arguments as a generic type and then extend them. That way we will never be missing the args in any of the functions that require these 3. See AI example.

```
type PushConfigResult = /* your PushConfigResult type */;
type PushKeyConfigResult = /* your PushKeyConfigResult type */;

type BaseConfig<T> = {
  groupInfo: T | null;
  groupMember: T | null;
  groupKeys: T | null;
};

type ExtendedConfig = BaseConfig<PushConfigResult>;

// Usage example:
const config: ExtendedConfig = {
  groupInfo: /* PushConfigResult value */,
  groupMember: null, // or a PushConfigResult value
  groupKeys: null, // or a PushConfigResult value
};
```

yougotwill replied:
Just saw you have a `Nullable` type which you could use with my idea

Bilb replied:
idk if that's worth it. the usages are not 100% the same.
```ts
 groupInfo: PushConfigResult | null;
        groupMember: PushConfigResult | null;
        groupKeys: PushKeyConfigResult | null;
```
and 
```
groupInfo: Array<MergeSingle> | null;
        groupMember: Array<MergeSingle> | null;
        groupKeys: Array<MergeSingle & { timestampMs: number }> | null;
        ```

yougotwill replied:
Talked about it

yougotwill commented on types/groups/metagroup.d.ts:
```diff
@@ -0,0 +1,129 @@
+    | MakeActionCall<MetaGroupWrapper, 'loadKeyMessage'>
+    | MakeActionCall<MetaGroupWrapper, 'currentHashes'>
+    | MakeActionCall<MetaGroupWrapper, 'encryptMessage'>
+    | MakeActionCall<MetaGroupWrapper, 'decryptMessage'>;
```
Very sexy 💅

yougotwill commented on types/shared.d.ts:
```diff
@@ -0,0 +1,152 @@
+    : never;
+  type _FixedSizeArray<T, N extends number, R extends unknown[]> = R['length'] extends N
+    ? R
+    : _FixedSizeArray<T, N, [T, ...R]>;
```
Damn some real all over the place 🧙

Bilb replied:
haha

yougotwill commented on types/shared.d.ts:
```diff
@@ -0,0 +1,152 @@
+declare module 'libsession_util_nodejs' {
+  type FixedSizeArray<T, N extends number> = N extends N
```
Make a note that it shouldn't be used for big uint arrays because it will cause the linter to poop itself

Bilb replied:
added a 
```ts
  /**
   * NOTE: For linting performance reason, this type should only be used with small fixed size arrays (less < 100 items)
   */
```

yougotwill commented on types/user/userconfig.d.ts:
```diff
@@ -11,13 +13,13 @@ declare module 'libsession_util_nodejs' {
       priority: number;
       url: string;
-      key: Uint8Array;
+      key: Uint8Array; // FixedSizeUint8Array<32>
```
Can you not use your new types here?

Bilb replied:
that type is actually not very convenient. I might need to tweak it. Like it's convenient to use while only using FixedSize arrays, but getting one from a Uint8array is annoying and getting the buffer from a FixedSizeUint9Array is annoying too

Bilb replied:
I'll keep this opened until I change it.

yougotwill commented on types/user/usergroups.d.ts:
```diff
@@ -1,39 +1,57 @@
+   * We can set anything on a UserGroup and can omit fields by explicitely setting them to null.
+   * The only one which cannot be omited is the pubkeyHex
+   */
+  type UserGroupsSet = Pick<UserGroupsGet, 'pubkeyHex'> &
```
🧙

Bilb committed:
chore: address pr reviews

Bilb committed:
chore: remove the fixedlength uint8array as it is not convenient at all

we will need a better way to do this, but also have nice
wrapping/unwrapping in the nodejs wrapper

Bilb committed:
feat: add subaccount calls

Bilb commented:
closing this in favor of https://github.com/oxen-io/libsession-util-nodejs/pull/13 (name of chunk 1 in the name, but there is nothing else to discuss here)
